### PR TITLE
Allow dependencies without a package.json

### PR DESCRIPTION
**Pull request by boardend**
Work in progress: A possible fix for #1866 (and dup #3728)

### CODE REVIEW

File: packages/git-fetcher/src/index.ts [hunk 1]
```diff
@@ -20,9 +21,13 @@ export default () => {
       const tempLocation = await cafs.tempDir()
       await execGit(['clone', resolution.repo, tempLocation])
       await execGit(['checkout', resolution.commit], { cwd: tempLocation })
-      await preparePackage(tempLocation)
       // removing /.git to make directory integrity calculation faster
       await rimraf(path.join(tempLocation, '.git'))
+      // invoke 'prepare' if package.json exists and defines a 'prepare' script
+      if (await exists(path.join(tempLocation, 'package.json'))) {
+        await preparePackage(tempLocation)
+      }
```
Comment (zkochan) [inline packages/git-fetcher/src/index.ts]: Why was this moved down from line 23?

File: packages/package-requester/src/packageRequester.ts [hunk 2]
```diff
@@ -234,7 +234,7 @@ async function resolveAndFetch (
   }
 
   const fetchResult = ctx.fetchPackageToStore({
-    fetchRawManifest: true,
+    fetchRawManifest: !!manifest,
```
Comment (zkochan) [inline packages/package-requester/src/packageRequester.ts]:
```suggestion
    fetchRawManifest: Boolean(manifest),
```

File: packages/git-fetcher/src/index.ts (hunk 1 above)
Comment (boardend) [reply packages/git-fetcher/src/index.ts]: I don't know :) I will move it back

Comment (boardend) [reply packages/git-fetcher/src/index.ts]: The lines are now in the original order. If you want, I could squash all commmits together? Otherwise it should be ready to merge.

**Comment by boardend** [comment]
I’m still not sure about the failing installation/workflows. If I add the new fixture to https://github.com/pnpm/pnpm/blob/main/fixtures/pnpm-workspace.yaml, the new test cases will fail. If I don’t add it, the Github workflows will fail during installation.

**Comment by zkochan** [comment]
I'll check

**Comment by zkochan** [comment]
I updated the script to use the pnpm from the repository to install dependencies for the fixtures. But it still fails.

**Comment by boardend** [comment]
Did a rebase to the current main, had to switch to `safeReadPackageFromDir` in `runPostinstallHooks` (see 26ec0cea9373d2b0f8eef01664c4972545299e7a) and also fix the path to `../../../packages/pnpm/dist/pnpm.cjs` after `cd ./fixture-with-external-shrinkwrap/pkg` in `fixtures/package.yaml` (see e43eb2d4cb433af797a97703be410ba618962455)

Now we should be ready to merge, hopefully :)

**Comment by boardend** [comment]
@zkochan do you have any idea what went wrong?

File: packages/package-requester/src/packageRequester.ts [hunk 3]
```diff
@@ -256,7 +256,12 @@ async function resolveAndFetch (
       resolvedVia,
       updated,
     },
-    bundledManifest: fetchResult.bundledManifest,
+    bundledManifest: manifest
+      ? fetchResult.bundledManifest
+      : () => Promise.resolve({
+        name: wantedDependency.alias,
+        version: wantedDependency.pref,
+      } as DependencyManifest),
```
Comment (zkochan) [inline packages/package-requester/src/packageRequester.ts]: the tests in `@pnpm/package-requester` fail because of this change. I don't think this package should be changed at all.
